docs/hugo/content/design/ADR-2022-11-Resource-Import.md

```diff
@@ -0,0 +1,78 @@
+---
+title: '2022-11: ASO v1 Migration and Resource Import'
+---
+
+## Context
+
+We have a significant number of customers successfully using ASO v1 whom we want to migrate to ASO v2. This requires both feature parity (so that users don't lose functionality when they switch) and a straightforward migration path (if it's too hard to migrate, they won't).
+
+Discussions so far have identified two possible solutions - *version integration* and *resource import*.
+### Version Integration
+
+We would add new hand-written resources into ASO v2 with the same names and shapes as the resources supported by ASO v1. Users would be able to migrate to ASO v2 simply by removing ASO v1 from their cluster and installing ASO v2 in its place.
+
+Since many of these resources are already supported by ASO v2, albeit with shapes much more closely aligned with the ARM originals, we would need to implement forward and backward conversions between the two shapes, preserving the Kubernetes version compatibility guarantees we have today.
+
+* Pro: Very simple migration for our customers.
+* Con: Would require a lot of work to implement and test the conversions.
+* Con: Any resource features that don't map exactly to a property in the ARM shape would be difficult to retain.
+* Con: The investment would benefit only a small number of customers.
+* Con: The conversions would need to be maintained, potentially indefinitely.
+* Con: It's unlikely we'd achieve 100% fidelity, requiring customers to change their YAML files anyway.
+
+### Resource Migration
+
+Create a tool that allowed any supported Azure resource to be to be exported as a YAML file in the shape expected by ASO v2. Users would then be able to modify the YAML file as necessary and apply it to their cluster.
+
+The most likely form of this tool would be as a command-line utility for users to run locally. Idiomatically for the Kubernetes ecosystem, we'd call this `asoctl`.
+
+* Pro: Straightforward migration for ASO v1 customers (though, not zero touch).
+* Pro: Would benefit all customers wanting to migrate existing Azure resources, not just those using ASO v1.
+* Pro: Could also be used to snapshot existing resources (e.g. to capture a hand configured resource as YAML for reuse).
+* Con: We'd need to extend the code generator to provide the required support (but much of what we need already exists).
+
+## Decision
+
+Create a new command line tool, `asoctl` to house multiple functions, starting with resource import. The tool will follow the usual convention of supporting a variety of verbs, each activating a different mode.
```

Comment:
I do still think we should include the "do it with a CRD" option and pros/cons for why we decided not to do it.

We can't just blindly apply the YAML we import as there are things we need the user to provide us (secrets being the main one we can't do without). There are many resources which don't have secrets though (or have secrets but which can be used w/ AAD authentication modes where they don't have secrets) and for those resources a direct import would work.

Amazon's offering does it this way: [ACK Adoption](https://aws-controllers-k8s.github.io/community/docs/user-docs/adopted-resource/)

Google's offering does it with a cmdline tool like you've proposed here: [ConfigConnector Export](https://cloud.google.com/config-connector/docs/how-to/import-export/export)

There are pros and cons to both so I'm not saying we should use the CRD approach but I think we should document it and say why we didn't.

The arguments I see are:
Pro: You don't need to use an `asoctl` tool, which means you can import/export using just `kubectl`.
Con: Awkward to model a run-once action as a CRD as there is no reconcile to be done after the first pass. There are examples of such resources in core k8s API such as `Job` and a pattern we could follow
Con: Would need to automatically mark resources as `skip` reconcile for safety, but then users need to go remove that which takes more work.

There may be other pros/cons, but given at least one of our competitors does it that way it seems like we should describe why we don't.

Reply:
Done.